### PR TITLE
Swift: route compiler diagnostics through our log

### PR DESCRIPTION
**Pull request by redsun82**

### CODE REVIEW

File: swift/extractor/invocation/SwiftDiagnosticsConsumer.cpp
Comment (redsun82) [inline swift/extractor/invocation/SwiftDiagnosticsConsumer.cpp:60]: There is a potential foot gun lurking here: one might be tempted to call `LOG_IMPL` directly with something like `LOG_IMPL(getLoggingLevelFromDiag(diagInfo.Kind), ...)`: that will happily compile, but because of how binlog works, the severity will be fixed to the first encountered one (severity is saved into a static variable).

I'm considering coding `LOG_IMPL` in a way that will give a compiler error if called with a non-constant severity, something like assigning the provided log level to a constexpr dummy variable.

Comment (redsun82) [reply swift/extractor/invocation/SwiftDiagnosticsConsumer.cpp:60]: I've done that in the other PR: https://github.com/github/codeql/tree/6ef908807651851c06dabe579e68e7e088c10cd3

**Review by sashabu — APPROVED** [review]
LGTM, but please hold off on merging until the base PR is merged (or at least ready to merge).

File: swift/extractor/main.cpp
Comment (AlexDenisov) [inline swift/extractor/main.cpp:90]: Does it mean we won't see the messages in the tracer log?

Comment (redsun82) [reply swift/extractor/main.cpp:90]: here's an example from running `swift/ql/test/extractor-tests/errors`:

```
2023-04-18 07:03:35.999366719 ERRO [extractor/compiler] unresolved.swift:3:10 generic parameter 'T' is not used in function signature (SwiftDiagnosticsConsumer.cpp:45)
2023-04-18 07:03:35.999527236 ERRO [extractor/compiler] unresolved.swift:5:1 cannot explicitly specialize a generic function (SwiftDiagnosticsConsumer.cpp:45)
```

Comment (AlexDenisov) [reply swift/extractor/main.cpp:90]: One of the concerts I have is that now we have one more layer of indirection when debugging things. I don't have enough experience to judge how bad/good this is, but this is something to keep an eye on.

Comment (redsun82) [reply swift/extractor/main.cpp:90]: as far as I can tell all the relevant info is still there in the tracer build log, and the conciseness (one line per error without the code snippet) actually helps readability, especially once all extractor logs will follow the same format. Typically the error description and the source line are enough to locate the issue. We can reiterate on this though if it proves to be a problem.

**Pull request merged**
